src/invert/laplace/impls/iterative_parallel_tri/iterative_parallel_tri.cxx

```diff
@@ -1576,22 +1468,38 @@ void LaplaceIPT::refine(Level &l, Level &lup, Matrix<dcomplex> &fine_error, cons
  * field(1,:) is correct, and puts the in-neighbour's value into field(0,:)
  * and out-neighbour's value into field(3,:).
  */
-void LaplaceIPT::synchronize_reduced_field(const Level &l, Matrix<dcomplex> &field){
+void LaplaceIPT::Level::synchronize_reduced_field(const LaplaceIPT& l,
+                                                  Matrix<dcomplex>& field) {
 
   SCOREP0();
-  if(l.included){
+  if (not included) {
+    return;
+  }
 
-    MPI_Comm comm = BoutComm::get();
-    int err;
+  MPI_Comm comm = BoutComm::get();
 
-    // Communicate in
-    if(!localmesh->firstX()){
-      err = MPI_Sendrecv(&field(1,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_in, 1, &field(0,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_in, 0, comm, MPI_STATUS_IGNORE);
-    }
+  // Communicate in
+  if (not l.localmesh->firstX()) {
+    MPI_Sendrecv(&field(1, 0), l.nmode, MPI_DOUBLE_COMPLEX, proc_in, 1, &field(0, 0),
+                 l.nmode, MPI_DOUBLE_COMPLEX, proc_in, 0, comm, MPI_STATUS_IGNORE);
+  }
 
-    // Communicate out
-    if(!localmesh->lastX()){
-      err = MPI_Sendrecv(&field(1,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_out, 0, &field(3,0), nmode, MPI_DOUBLE_COMPLEX, l.proc_out, 1, comm, MPI_STATUS_IGNORE);
+  // Communicate out
+  if (not l.localmesh->lastX()) {
+    MPI_Sendrecv(&field(1, 0), l.nmode, MPI_DOUBLE_COMPLEX, proc_out, 0, &field(3, 0),
+                 l.nmode, MPI_DOUBLE_COMPLEX, proc_out, 1, comm, MPI_STATUS_IGNORE);
+  }
+}
+
+/*
+ * Returns the transpose of a matrix
+ */
+void LaplaceIPT::transpose(Matrix<dcomplex>& m_t, const Matrix<dcomplex>& m, const int n1,
+                           const int n2) {
+  SCOREP0();
+  for (int i1 = 0; i1 < n1; i1++) {
+    for (int i2 = 0; i2 < n2; i2++) {
+      m_t(i1, i2) = m(i2, i1);
```

Comment:
I'd prefer this to return `m_t` instead of having it as an out-argument, but we don't currently have a move-constructor on `Matrix` that would make this efficient.

We can avoid needing to pass in `n1, n2` though:

```suggestion
void LaplaceIPT::transpose(Matrix<dcomplex>& m_t, const Matrix<dcomplex>& m) {
  SCOREP0();
  const auto n1 = std::get<0>(m.shape());
  const auto n2 = std::get<1>(m.shape());
  for (int i1 = 0; i1 < n1; i1++) {
    for (int i2 = 0; i2 < n2; i2++) {
      m_t(i1, i2) = m(i2, i1);
```

Reply:
Done